File: Protocol/Protocol/Protocol/Car/Car.m
//
//  Car.m
//  Protocol
//
//  Created by 김기현 on 2021/02/16.
//

#import <Foundation/Foundation.h>
#import "Car.h"
#import "Tire.h"
#import "Engine.h"

@implementation Car

@synthesize name = appellation;
@synthesize engine;

// MARK: - Initializer
- (id) init {
    if (self = [super init]) {
        self.name = @"Car";
        tires = [[NSMutableArray alloc] init];
        
        int i;
        for (i = 0; i < 4; i++) {
            [tires addObject: [NSNull null]];
        }
    }
    
    return self;
} // init

- (id) copyWithZone: (NSZone *) zone {
    Car * carCopy;
    carCopy = [[[self class] allocWithZone: zone] init];
    
    carCopy.name = self.name;
    
    Engine * engineCopy;
    engineCopy = [engine copy];
    carCopy.engine = engineCopy;
    
    int i;
    for (i = 0; i < 4; i++) {
        Tire * tireCopy;
        tireCopy = [[self tireAtIndex: i] copy];
        
        [carCopy setTire: tireCopy atIndex: i];
    }
    
    return carCopy;
}

// MARK: - Method
- (void) setTire: (Tire *) tire atIndex: (int) index {
    [tires replaceObjectAtIndex: index withObject: tire];
} // setTire:atIndex:

- (Tire *) tireAtIndex: (int) index {
    Tire * tire;
    tire = [tires objectAtIndex: index];
    
    return tire;
} // tireAtIndex:

- (void) print {
    NSLog(@"%@ has : ", self.name);
    
    int i;
    for (i = 0; i < 4; i++) {
        NSLog(@"%@", [self tireAtIndex: i]);
    }
    
    NSLog(@"%@", engine);
} // print

@end // Car

File: CarProtocol Swift/CarProtocolSwift/CarProtocolSwift/Car/Car.swift
//
//  Car.swift
//  CarProtocolSwift
//
//  Created by 김기현 on 2021/02/17.
//

import Foundation

class Car: NSCopying {
    var appellation: String?
    var tires: [Tire]?
    
    var name: String?
    var engine: Engine?
    
    required init() {
        name = "Car"
        tires = [Tire]()
        
        for _ in 0..<4 {
            tires?.append(Tire())
        }
    }
    
    func copy(with zone: NSZone? = nil) -> Any {
        let carCopy: Car?
        carCopy = type(of: self).init()
        
        guard let result = carCopy else { fatalError("car copy fail") }
        return result
    }
    
    func setTire(_ tire: Tire?, at index: Int) {
        guard var tires = self.tires, let tire = tire else { fatalError("tires or tire nil") }
        tires[index] = tire
    }
    
    func tire(at index: Int) -> Tire? {
        let tire: Tire?
        
        guard let tires = self.tires else { fatalError("tires nil") }
        tire = tires[index]
        
        return tire
    }
    
    func printLog() {
        guard let name = self.name,
              let engine = self.engine else { fatalError() }
        
        print("\(String(describing: name)) has : ")
        
        for i in 0..<4 {
            guard let result = tire(at: i) else { fatalError() }
            print("\(result)")
        }
        
        print("\(String(describing: engine))")
    }
}
